test/helpers/kubectl.go

```diff
@@ -626,41 +626,79 @@ func (kub *Kubectl) WaitKubeDNS() error {
 
-	if !strings.HasSuffix(name, svcSuffix) {
-		name = fmt.Sprintf("%s.%s", name, svcSuffix)
+	serviceNameWithNamespace := fmt.Sprintf("%s.%s", serviceName, serviceNamespace)
+	if !strings.HasSuffix(serviceNameWithNamespace, svcSuffix) {
+		serviceNameWithNamespace = fmt.Sprintf("%s.%s", serviceNameWithNamespace, svcSuffix)
 	}
 	// https://bugs.launchpad.net/ubuntu/+source/bind9/+bug/854705
 	digCMD := "dig +short %s @%s | grep -v -e '^;'"
 
+	// Also provide full dig output for debugging purposes.
+	digCmdLong := "dig %s @%s"
+
 	// If it fails we want to know if it's because of connection cannot be
 	// established or DNS does not exist.
 	digCMDFallback := "dig +tcp %s @%s"
 
-	host, _, err := kub.GetServiceHostPort(KubeSystemNamespace, "kube-dns")
+	dnsClusterIP, _, err := kub.GetServiceHostPort(KubeSystemNamespace, "kube-dns")
 	if err != nil {
 		logger.WithError(err).Error("cannot get kube-dns service IP")
 		return err
 	}
 
 	body := func() bool {
-		res := kub.Exec(fmt.Sprintf(digCMD, name, host))
+		serviceIP, _, err := kub.GetServiceHostPort(serviceNamespace, serviceName)
+		if err != nil {
+			log.WithError(err).Errorf("cannot get service IP for service %s", serviceNameWithNamespace)
+			return false
+		}
+
+		// ClusterIPNone denotes that this service is headless; there is no
+		// service IP for this service, and thus the IP returned by `dig` is
+		// an IP of the pod itself, not ClusterIPNone.
+		if serviceIP != v1.ClusterIPNone {
+			log.Debugf("service is not headless; checking whether IP retrieved from DNS matches the IP for the service stored in Kubernetes")
+			res := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
+			serviceIPFromDNS := res.SingleOut()
+			if !govalidator.IsIP(serviceIPFromDNS) {
+				logger.Debugf("output of dig (%s) did not return an IP")
+				return false
+			}
+
+			// Due to lag between new IPs for the same service being synced between
+			// kube-apiserver and DNS, check if the IP for the service that is
+			// stored in K8s matches the IP of the service cached in DNS. These
+			// can be different, because some tests use the same service names.
+			// Wait accordingly for services to match, and for resolving the service
+			// name to resolve via DNS.
+			if !strings.Contains(serviceIPFromDNS, serviceIP) {
+				logger.Debugf("service IP retrieved from DNS (%s) does not match the IP for the service stored in Kubernetes (%s)", serviceIPFromDNS, serviceIP)
+				_ = kub.Exec(fmt.Sprintf(digCmdLong, serviceNameWithNamespace, dnsClusterIP))
+				_ = kub.Exec(fmt.Sprintf(digCMDFallback, serviceNameWithNamespace, dnsClusterIP))
+				return false
+			}
+			logger.Debugf("service IP retrieved from DNS (%s) matches the IP for the service stored in Kubernetes (%s)", serviceIPFromDNS, serviceIP)
+		}
+
+		res := kub.Exec(fmt.Sprintf(digCMD, serviceNameWithNamespace, dnsClusterIP))
+		_ = kub.Exec(fmt.Sprintf(digCmdLong, serviceNameWithNamespace, dnsClusterIP))
```

Comment:
I don't think that the `digCMDLog` is needed at all. The full output is with the digCMDFallback, if you use here again there is no value added at all. I would delete it.

Reply:
OK; the reason I added `digCmdLog` was for when `dig` succeeded (an IP was found), but it wasn't the right IP for the service. We can just unconditionally run `digCmdFallback` in that case.